src/DynamoCoreWpf/Utilities/ActionDebouncer.cs

```diff
@@ -0,0 +1,45 @@
+using System;
+using System.Threading;
+using System.Windows.Threading;
+
+namespace Dynamo.Wpf.Utilities
+{
+    internal class ActionDebouncer : IDisposable
+    {
+        private readonly Dispatcher dispatcher;
+        private CancellationTokenSource cts;
+
+        public ActionDebouncer(Dispatcher dispatcher)
+        {
+            this.dispatcher = dispatcher;
+        }
+
+        public void Cancel()
+        {
+            if (cts != null)
+            {
+                cts.Cancel();
+                cts.Dispose();
+                cts = null;
+            }
+        }
+
+        public void Debounce(int timeout, Action action)
+        {
+            Cancel();
+            cts = new CancellationTokenSource();
+            System.Threading.Tasks.Task.Delay(timeout, cts.Token).ContinueWith(t =>
+            {
+                if (t.IsCompletedSuccessfully)
+                {
+                    dispatcher.BeginInvoke(action);
+                }
+            });
+        }
```

Comment:
If you simply assume (document) that this class should only be called on the UI thread (not enforced), you could simply schedule the continuation on the same thread by using the `TaskContinuationOptions.ExecuteSynchronously` option.
This way you do not even need the dispatcher.
Passing a generic dispatcher to the constructor does not really enforce it being use in the UI thread anyway.

Reply:
that sounds like a really good simplification.